Approved/enhanced-array-literals.md

```diff
@@ -0,0 +1,251 @@
+let wrong1 = [size = 10, false];
+let wrong2 = [false, true, size = 10];
+let wrong3 = [false, size = 10, true];
+```
+
+# Implementation
+
+The implementation of `[x, size = n]` should be a relatively straightforward extension of the current implementation for `new T[n]`.
+It will require parser and syntax tree changes, as well as changes to code generation and potentially a small amount of runtime support.
+
+The implementation for `[]` will require enhancements to type inference.
+One possible approach is to create a placeholder type variable for each usage of `[]`.
+When an expression with a placeholder type variable is used in a context that requires a specific type, this context can refine the placeholder type.
+Once all placeholders are refined, they can be replaced with their concrete types if one was determined.
+Since all specialization signatures must have explicit parameter and return types, the type inference for a placeholder variable cannot extend beyond its containing specialization.
+
+## Timeline
+
+The bulk of the work is expected to be in implementing type inference for the empty array literal.
+However, since it is limited to only the empty array literal, and bounded by the scope of a specialization, the complexity should be managable within a month or two of work.
+The remaining work of adding repeated constant array literals and deprecating the old array literals should not take much time.
+
+# Further Considerations
+
+## Related Mechanisms
+
+The syntax for allocating arrays of qubits is similar to the existing default-initialized array constructor syntax:
+
+```qsharp
+use qs = Qubit[n];
+```
+
+For consistency, it may make sense to change this syntax as well.
+The [proposal for allocatable types and generalization of initializers](https://github.com/microsoft/qsharp-language/pull/41) is addressing this.
+
+## Impact on Existing Mechanisms
+
+The [`Default`](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.core.default) function depends on default-initialized array constructors, and it will also be deprecated.
+
+## Anticipated Interactions with Future Modifications
+
+### Named arguments
+
+The syntax for repeated constant arrays is designed to be similar to the syntax for named arguments, a feature that Q# does not currently support (and for which there is currently no suggestion or proposal).
+This proposal makes the assumption that the syntax for named arguments, if Q# supports them, will be:
+
+```qsharp
+Foo(firstArgument, secondArgument, namedArgument = "foo");
+```
+
+### Type inference
+
+In the future, the type inference used for empty array literals may be extended to all expressions, allowing calls to polymorphic functions where not all type variables are determined by their arguments.
+
+### Partial application
+
+We may want to consider supporting partial application for array literals in the future:
+
+```qsharp
+let f = ["foo", size = _];
+// f : Int -> String[]
```

Comment:
After thinking about this some more, I don't think the assertion that `f : Int -> String[]` is accurate in general. I think partial application of the `size` parameter is problematic. See the Raised Concerns section for more about this.

Reply:
Updated the proposal to clarify that a fixed type for `size` should be used.